flask_pymongo/__init__.py

```diff
@@ -237,5 +238,19 @@ def save_upload(filename):
             db_obj = self.db
         assert db_obj is not None, "Please initialize the app before calling save_file!"
         storage = GridFS(db_obj, base)
-        id = storage.put(fileobj, filename=filename, content_type=content_type, **kwargs)
+
+        # GridFS does not manage its own hash, so we manage our own using its
+        # metadata storage, to be used for the etag.
+        sha1_sum = self._compute_sha(fileobj)
+        metadata = dict(sha1_sum=sha1_sum)
+        id = storage.put(
+            fileobj, filename=filename, content_type=content_type, metadata=metadata, **kwargs
```

Comment:
Hmm I just had another thought. GridOut is seekable but fileobj on save_file() here can be any file-like object and IIRC not every such object is seekable, is that true?. If so we may need to upload the file and compute the hash in one pass.